src/emulator/cpu.ts

```diff
@@ -336,119 +336,130 @@ export class Cpu {
+                const index = this.bus.read16((this.state.p + 1) & 0xffff);
+                return this.bus.read(index);
+            }
+
+            case AddressingMode.imm8io: {
+                const index = this.bus.read((this.state.p + 1) & 0xffff);
+                return this.bus.read(0xff00 + index);
+            }
 
             case AddressingMode.reg8:
-            case AddressingMode.reg8_imm8:
-            case AddressingMode.reg8_ind8:
-            case AddressingMode.reg8_reg8:
-            case AddressingMode.reg8_reg8io:
-            case AddressingMode.reg8_imm8io:
-                return this.state.r8[instruction.par1];
-
-            case AddressingMode.reg16_imm16:
-            case AddressingMode.reg16:
-                return this.state.r16[instruction.par1];
+                if (par.value === undefined) {
+                    throw new Error('missing value for operation parameter');
+                }
+
+                return this.state.r8[par.value];
+
+            case AddressingMode.reg8io:
+                if (par.value === undefined) {
+                    throw new Error('missing value for operation parameter');
+                }
+
+                return this.bus.read(0xff00 + this.state.r8[par.value]);
 
-            case AddressingMode.ind8_reg8:
-            case AddressingMode.ind8_imm8:
             case AddressingMode.ind8:
-                return this.bus.read(this.state.r16[instruction.par1]);
+                if (par.value === undefined) {
+                    throw new Error('missing value for operation parameter');
+                }
 
-            case AddressingMode.reg8io_reg8:
-                return this.bus.read(0xff00 + this.state.r8[instruction.par1]);
+                return this.bus.read(this.state.r16[par.value]);
 
-            case AddressingMode.imm8io_reg8: {
-                const index = this.bus.read((this.state.p + 1) & 0xffff);
-                return this.bus.read(0xff00 + index);
-            }
+            case AddressingMode.imm16:
+                return this.bus.read16((this.state.p + 1) & 0xffff);
+
+            case AddressingMode.reg16:
+                if (par.value === undefined) {
+                    throw new Error('missing value for operation parameter');
+                }
+
+                return this.state.r16[par.value];
 
             default:
                 throw new Error('bad addressing mode');
         }
     }
 
-    private setArg1(instruction: Instruction, value: number) {
-        switch (instruction.addressingMode) {
-            case AddressingMode.reg8:
-            case AddressingMode.reg8_imm8:
-            case AddressingMode.reg8_ind8:
-            case AddressingMode.reg8_reg8:
-            case AddressingMode.reg8_imm8io:
-            case AddressingMode.reg8_reg8io:
-                this.state.r8[instruction.par1] = value & 0xff;
-                break;
+    private getArg1(instruction: Instruction): number {
```

Comment:
I would just get rid of this as it is a mere trampoline with an overhead that is negligible in the best case. The null check will be obsolete with my other changes suggested below.

Reply:
On second thought: let's keep it (without the check) --- keeps things readable and should be optimized out by the engine.